zjit/src/hir.rs

```diff
@@ -3677,28 +3362,22 @@ impl Function {
             let mut new_insns = vec![];
             for insn_id in old_insns {
                 let replacement_id = match self.find(insn_id) {
-                    Insn::GuardType { val, guard_type, .. } if self.is_a(val, guard_type) => {
-                        self.make_equal_to(insn_id, val);
-                        // Don't bother re-inferring the type of val; we already know it.
-                        continue;
-                    }
-                    Insn::LoadField { recv, offset, return_type, .. } if return_type.is_subtype(types::BasicObject) &&
-                            u32::try_from(offset).is_ok() => {
-                        let offset = (offset as u32).to_usize();
-                        let recv_type = self.type_of(recv);
-                        match recv_type.ruby_object() {
-                            Some(recv_obj) if recv_obj.is_frozen() => {
-                                let recv_ptr = recv_obj.as_ptr() as *const VALUE;
-                                let val = unsafe { recv_ptr.byte_add(offset).read() };
-                                self.new_insn(Insn::Const { val: Const::Value(val) })
+                    Insn::GuardType { val, guard_type, .. } => {
+                        // If we already know the type matches, eliminate the guard
+                        if self.is_a(val, guard_type) {
+                            self.make_equal_to(insn_id, val);
+                            continue;
+                        }
+                        // If GuardType is on a frozen object, we can check at compile time if the
```

Comment:
I think this should be implied by the `is_a` in the existing code; we can generally infer types from constant objects (VALUE pointers). Ideally we instead eliminate GuardShape instructions, which we don't have yet